core/src/template-string/ast.ts

```diff
@@ -0,0 +1,640 @@
+    })
+  }
+}
+
+export class EndIfBlockExpression extends TemplateExpression {
+  override evaluate(): never {
+    // See also `buildConditionalTree` in `parser.pegjs`
+    throw new InternalError({
+      message: `{endif} block expression should not end up in the final AST`,
+    })
+  }
+}
+
+export class IfBlockExpression extends TemplateExpression {
+  constructor(
+    loc: Location,
+    public readonly condition: TemplateExpression,
+    public ifTrue: TemplateExpression,
+    public ifFalse: TemplateExpression | undefined
+  ) {
+    super(loc)
+  }
+
+  override evaluate(args: EvaluateArgs): CollectionOrValue<TemplateLeaf> {
+    const condition = this.condition.evaluate(args)
+
+    const evaluated = isTruthy(condition) ? this.ifTrue.evaluate(args) : this.ifFalse?.evaluate(args)
+
+    return mergeInputs(
+      evaluated ||
+        new TemplateLeaf({
+          expr: args.rawTemplateString,
+          value: "",
+          inputs: {},
+        }),
+      condition
+    )
+  }
+}
+
+export class StringConcatExpression extends TemplateExpression {
+  public readonly expressions: TemplateExpression[]
+  constructor(loc: Location, ...expressions: TemplateExpression[]) {
+    super(loc)
+    this.expressions = expressions
+  }
+
+  override evaluate(args: EvaluateArgs): TemplateLeaf<string> {
+    const evaluatedExpressions: TemplateLeaf<TemplatePrimitive>[] = this.expressions.map((expr) => {
+      const r = expr.evaluate(args)
+
+      if (!isTemplateLeaf(r) || !isTemplatePrimitive(r.value)) {
+        throw new TemplateStringError({
+          message: `Cannot concatenate: expected primitive, but expression resolved to ${
+            isTemplateLeaf(r) ? typeof r.value : typeof r
+          }`,
+        })
+      }
+
+      // The isPrimitive asserts that we are dealing with primitive values, and not empty arrays
+      return r as TemplateLeaf<TemplatePrimitive>
+    })
+
+    const result = evaluatedExpressions.reduce((acc, expr) => {
+      return `${acc}${expr.value === undefined ? "" : expr.value}`
+    }, "")
+
+    return mergeInputs(
+      new TemplateLeaf({
+        expr: args.rawTemplateString,
+        value: result,
+        inputs: {},
+      }),
+      ...evaluatedExpressions
+    ) as TemplateLeaf<string> // TODO: fix mergeInputs return type
+  }
+}
+
+export class MemberExpression extends TemplateExpression {
+  constructor(
+    loc: Location,
+    public readonly innerExpression: TemplateExpression
+  ) {
+    super(loc)
+  }
+
+  override evaluate(args: EvaluateArgs): TemplateLeaf<string | number> {
+    const inner = this.innerExpression.evaluate(args)
+    const innerValue = isTemplateLeaf(inner) ? inner.value : inner
+
+    if (typeof innerValue !== "string" && typeof innerValue !== "number") {
+      throw new TemplateStringError({
+        message: `Expression in bracket must resolve to a string or number (got ${typeof innerValue}).`,
+      })
+    }
+
+    return new TemplateLeaf({
```

Comment:
This wipes the inputs of the inner expression;
We need to actually:

```suggestion
    return inner
```